daemon: fix under what conditions container's mac-address is applied

akerouanton opened the pull request:
**- What I did**

- Fix https://github.com/moby/moby/issues/46404
- Fix docker/compose#10796

The daemon would pass an EndpointCreateOption to set the interface MAC address if the network name and the provided network mode were matching. Obviously, if the network mode is a network ID, it won't work. To make things worse, the network mode is never normalized if it's a partial ID.

To fix that: 1. the condition under what the container's mac-address is applied is updated to also match the full ID; 2. the network mode is normalized to a full ID when it's only a partial one.

**- How to verify it**

```
$ docker run --rm --name test --network 6cabee8406 --mac-address "02:42:ac:11:00:02" nicolaka/netshoot ip -brief link show dev eth0
eth0@if92        UP             02:42:ac:11:00:02 <BROADCAST,MULTICAST,UP,LOWER_UP> 
```

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://upload.wikimedia.org/wikipedia/commons/d/d4/Sea_otter_pair2.jpg)

akerouanton commented on daemon/network.go:
```diff
@@ -866,20 +866,18 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
-			}))
+	if c.Config.MacAddress != "" {
+		if c.HostConfig.NetworkMode.IsHost() || c.HostConfig.NetworkMode.IsNone() || c.HostConfig.NetworkMode.IsContainer() {
+			return nil, fmt.Errorf("you can't specify a mac-address for a container with network-mode %s", c.HostConfig.NetworkMode)
```
I realize the daemon logs an extra `FIXME` when this error is returned because it's not a proper `InvalidParameter`.

thaJeztah commented on daemon/network.go:
```diff
@@ -866,20 +866,18 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 		createOptions = append(createOptions, libnetwork.CreateOptionDisableResolution())
 	}
 
-	// configs that are applicable only for the endpoint in the network
-	// to which container was connected to on docker run.
-	// Ideally all these network-specific endpoint configurations must be moved under
-	// container.NetworkSettings.Networks[n.Name()]
-	if nwName == c.HostConfig.NetworkMode.NetworkName() || (nwName == defaultNetName && c.HostConfig.NetworkMode.IsDefault()) {
-		if c.Config.MacAddress != "" {
```
So the `c.Config.MacAddress` option is a "global" option, and applies to the container's main / primary network, and not a per-network options.

Does this change means that it applies to "any network" attached to the container? What happens if the container would be attached to multiple networks; do all endpoints get the same MacAddress?

akerouanton committed:
daemon: fix under what conditions container's mac-address is applied

The daemon would pass an EndpointCreateOption to set the interface MAC
address if the network name and the provided network mode were matching.
Obviously, if the network mode is a network ID, it won't work. To make
things worse, the network mode is never normalized if it's a partial ID.

To fix that: 1. the condition under what the container's mac-address is
applied is updated to also match the full ID; 2. the network mode is
normalized to a full ID when it's only a partial one.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton commented on daemon/network.go:
```diff
@@ -870,7 +870,8 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 	// container.NetworkSettings.Networks[n.Name()]
-	if nwName == c.HostConfig.NetworkMode.NetworkName() || (nwName == defaultNetName && c.HostConfig.NetworkMode.IsDefault()) {
+	netMode := c.HostConfig.NetworkMode
+	if nwName == netMode.NetworkName() || strings.HasPrefix(n.ID(), netMode.NetworkName()) || (nwName == defaultNetName && netMode.IsDefault()) {
```
Of course, this is still wrong... `netMode.NetworkName()` might be a partial prefix matching multiple things, for instance a network ID and a network name.

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@akerouanton looks like there's some (potentially) legit failures;

```
=== Failed
=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachable (1.22s)
    docker_cli_swarm_test.go:402: assertion failed: error is not nil: exit status 125: 8e83551cb95e7f06866b463de38a149694b8ede41d088691ee01a233c7d066ca
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode ovnet: network ovnet not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachable (1.22s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachableOnSwarmLeave (1.75s)
    docker_cli_swarm_test.go:421: assertion failed: error is not nil: exit status 125: 3ff1c5cde4b9564035c0399270b8a4fb67611378aa29b47832c827971ad50527
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode attovl: network attovl not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachableOnSwarmLeave (1.75s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestOverlayAttachableReleaseResourcesOnFailure (1.73s)
    docker_cli_swarm_test.go:446: assertion failed: error is not nil: exit status 125: 160bd7ef1111009c0dba21883f66ef6f698478d75b5403de5b6b384b4d71781f
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode ovnet: network ovnet not found.
        
    --- FAIL: TestDockerSwarmSuite/TestOverlayAttachableReleaseResourcesOnFailure (1.73s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerAttachByNetworkId (1.78s)
    docker_cli_swarm_test.go:370: assertion failed: error is not nil: exit status 125: be84a50271720d08a141fcee7127feaeca79cf28a131e8bb8e673e60a53f78d3
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode 6sux43f5ze7wjk74fcniahx9j: network 6sux43f5ze7wjk74fcniahx9j not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerAttachByNetworkId (1.78s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerAutoStart (1.75s)
    docker_cli_swarm_test.go:322: assertion failed: error is not nil: exit status 125: 9d2314700c35d5bdb113da738d03ca27505bbf6b212a0d94a258bc7a861ac919
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode foo: network foo not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerAutoStart (1.75s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmContainerEndpointOptions (4.33s)
    docker_cli_swarm_test.go:344: assertion failed: error is not nil: exit status 125: 25cad7271c2543866de6f9cb6aff7c7a01c16a0b23081de29b4ee73869125c8c
        Unable to find image 'busybox:glibc' locally
        glibc: Pulling from library/busybox
        3f4d90098f5b: Pulling fs layer
        3f4d90098f5b: Verifying Checksum
        3f4d90098f5b: Download complete
        3f4d90098f5b: Pull complete
        Digest: sha256:7338d0c72c655d6534670ed72590cdc99afc9c559958c47eb3a5a17a7520a75c
        Status: Downloaded newer image for busybox:glibc
        /usr/local/cli-integration/docker: Error response from daemon: could not find a network matching network mode foo: network foo not found.
        
    --- FAIL: TestDockerSwarmSuite/TestSwarmContainerEndpointOptions (4.33s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite (1441.65s)
```

thaJeztah reviewed (changes requested):
guess "not yet lgtm" then 😅

akerouanton commented on daemon/container_operations.go:
```diff
@@ -691,6 +691,12 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
 	}
 	nwName := n.Name()
 
+	if idOrName != container.HostConfig.NetworkMode.NetworkName() {
```
I really don't like the idea of mutating the network mode in `connectToNetwork` but there's no other choice: if the network mode references a swarm-scope network, it might not exist on the local node before `findAndAttachNetwork` is executed. As a Swarm agent doesn't have access to the list of swarm-scope networks, this is the only place where `normalizeNetMode` can be called.

As discussed with @thaJeztah today, I think it'd be preferable to create uninitialized/placeholder swarm-scope networks on every node. That would make the UX better and it'd restore the original semantics (ie. a container can't connect to a network that don't exist locally). That would make the code easier to reason about, we'd be able to move the call to `normalizeNetMode` upper in the call stack, and we'd not need to create a Swarm task in `findAndAttachNetwork` (ie. this too could be moved upper in the stack).

akerouanton commented on daemon/container_operations.go:
```diff
@@ -886,6 +892,25 @@ func (daemon *Daemon) tryDetachContainerFromClusterNetwork(network *libnetwork.N
 
+// normalizeNetMode checks whether the network mode references a network by a partial ID. In that case, it replaces the
+// partial ID with the full network ID.
+// TODO(aker): transform ID into name when the referenced network is one of the predefined.
```
This should be done in a follow-up PR.

thaJeztah reviewed (approved):
LGTM


sorry, thought we merged this one already, but I guess it had that one fix remaining, and I forgot 🙈

NonaSuomy commented:
Can't for the life of me get this working I tried to compile moby and replaced dockerd with the compile it still won't work with mac address
```
docker version
Client:
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.6
 Git commit:        ced0996600
 Built:             Wed Jul 26 21:44:58 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          dev
  API version:      1.44 (minimum version 1.12)
  Go version:       go1.21.1
  Git commit:       7d30dafd69ed06e94a0e1b87f74a1f7c9a98dfeb
  Built:            Sat Sep 30 17:52:34 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.6
  GitCommit:        091922f03c2762540fd057fba91260237ff86acb.m
 runc:
  Version:          1.1.9
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

```
# HomeAssistant Container for Docker
version: "3.9"
services:
  homeassistant:
    container_name: "homeassistant"
    image: "ghcr.io/home-assistant/home-assistant:beta"
    labels:
      - "com.centurylinklabs.watchtower.enable=true"
      - "com.centurylinklabs.watchtower.monitor-only=false"
    volumes:
      - /var/lib/libvirt/images/usb001/docker-storage/homeassistant/config:/config
      - /etc/localtime:/etc/localtime:ro
      - /run/dbus:/run/dbus:ro
    mac_address: "9e:de:ad:be:ef:23"
    hostname: "homeassistant"
    networks:
      - dhcp
    devices:
      - /dev/ttyACM0:/dev/ttyACM0
    restart: "no"
    privileged: true
    #network_mode: host

networks:
  dhcp:
    name: "dbrv100"
    external: true
```
Still grabs a random mac address.

NonaSuomy commented:
Hmm, now it's working after letting it settle for a while...
```
docker inspect network dbrv100
[
    {
        "Name": "dbrv100",
        "Id": "<networkid>",
        "Created": "2022-08-04T01:08:51.386572932-04:00",
        "Scope": "local",
        "Driver": "ghcr.io/devplayer0/docker-net-dhcp:release-linux-amd64",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "null",
            "Options": {},
            "Config": [
                {
                    "Subnet": "0.0.0.0/0"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "<containerID>": {
                "Name": "homeassistant",
                "EndpointID": "<endpointID>",
                "MacAddress": "9e:de:ad:be:ef:23",
                "IPv4Address": "10.0.10.100/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "bridge": "brv100"
        },
        "Labels": {}
    }
]
Error: No such object: network
```
